File: backend/src/main/java/com/example/skillsim/service/ScoreService.java
package com.example.skillsim.service;

import com.example.skillsim.config.ScoreDataLoader;
import com.example.skillsim.dto.ScoreRequest;
import com.example.skillsim.dto.ScoreResponse;
import com.example.skillsim.dto.ScoreSelection;
import com.example.skillsim.dto.ScoreSkillOption;
import com.example.skillsim.dto.ScoreTableRequest;
import com.example.skillsim.dto.ScoreTableResponse;
import com.example.skillsim.enums.Level;
import com.example.skillsim.model.ScoreEffect;
import com.example.skillsim.model.ScoreSkill;
import com.example.skillsim.repository.ScoreSkillRepository;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Comparator;
import java.util.EnumMap;
import java.util.Set;
import java.util.function.Supplier;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class ScoreService {

    private final ScoreSkillRepository scoreSkillRepository;
    private final ScoreCalculator scoreCalculator;
    private final Supplier<Map<String, Double>> statWeightsSupplier;

    @Autowired
    public ScoreService(
            ScoreSkillRepository scoreSkillRepository,
            ScoreCalculator scoreCalculator,
            ScoreDataLoader scoreDataLoader
    ) {
        this(scoreSkillRepository, scoreCalculator, scoreDataLoader::getStatWeights);
    }

    ScoreService(
            ScoreSkillRepository scoreSkillRepository,
            ScoreCalculator scoreCalculator,
            Map<String, Double> statWeights
    ) {
        this(scoreSkillRepository, scoreCalculator, () -> statWeights);
    }

    private ScoreService(
            ScoreSkillRepository scoreSkillRepository,
            ScoreCalculator scoreCalculator,
            Supplier<Map<String, Double>> statWeightsSupplier
    ) {
        this.scoreSkillRepository = scoreSkillRepository;
        this.scoreCalculator = scoreCalculator;
        this.statWeightsSupplier = statWeightsSupplier;
    }

    public List<ScoreSkillOption> listSkills(String cardType, String position) {
        String normalizedCardType = normalizeCardTypeOrThrow(cardType);
        String normalizedPosition = normalizeRequiredOrThrow(position, "Position selection is required.");

        return scoreSkillsForCardType(normalizedCardType).stream()
                .filter(skill -> SkillRules.matchesPosition(skill.getPosition(), normalizedPosition))
                .map(this::toOption)
                .toList();
    }

    public ScoreResponse calculate(ScoreRequest request) {
        if (request == null) {
            throw badRequest("Score request is required.");
        }

        String normalizedCardType = normalizeCardTypeOrThrow(request.getCardType());
        String normalizedPosition = normalizeRequiredOrThrow(request.getPosition(), "Position selection is required.");
        List<ScoreSelection> selections = request.getSelections();
        validateSelections(selections, normalizedCardType);

        String role = SkillRules.roleForPosition(normalizedPosition);
        Integer battingOrder = null;
        Integer pitcherSlot = request.getPitcherSlot();
        if ("BATTER".equals(role)) {
            battingOrder = validateBattingOrder(request.getBattingOrder());
        } else if ("SP".equals(role)) {
            if (pitcherSlot == null || pitcherSlot < 1 || pitcherSlot > 5) {
                throw badRequest("Pitcher slot must be between 1 and 5 for starting pitchers.");
            }
        } else if ("RP".equals(role)) {
            if (pitcherSlot == null || pitcherSlot < 1 || pitcherSlot > 6) {
                throw badRequest("Pitcher slot must be between 1 and 6 for relief pitchers.");
            }
        }
        Set<String> seenSkillIds = new HashSet<>();
        List<ScoreCalculator.Selection> calculatorSelections = new ArrayList<>();

        for (ScoreSelection selection : selections) {
            String skillId = normalizeRequiredOrThrow(selection.getSkillId(), "Skill selection is required.");
            if (!seenSkillIds.add(skillId)) {
                throw badRequest("Duplicate skill selection is not allowed.");
            }

            ScoreSkill skill = scoreSkillRepository.findBySkillKey(skillId)
                    .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Score skill not found: " + skillId));
            if (!allowedSkillCardTypes(normalizedCardType).contains(normalizeCardTypeOrThrow(skill.getCardType()))) {
                throw badRequest("Selected skill does not match requested card type.");
            }
            if (!SkillRules.matchesPosition(skill.getPosition(), normalizedPosition)) {
                throw badRequest("Selected skill does not match requested position.");
            }
            int maxLevel = maxLevel(skill);
            if (selection.getLevel() == null || selection.getLevel() < 1 || selection.getLevel() > maxLevel) {
                throw badRequest("Skill level must be between 1 and " + maxLevel + ".");
            }

            calculatorSelections.add(new ScoreCalculator.Selection(skill, selection.getLevel()));
        }

        Map<String, Double> conditionProbabilities = ScoreCalculator.conditionProbabilitiesForPosition(
                normalizedPosition,
                battingOrder,
                pitcherSlot,
                normalizedCardType,
                request.getThrowHand(),
                request.getBatHand()
        );
        UndefinedConditionWarnings undefinedConditionWarnings = applyUndefinedConditionWarnings(
                calculatorSelections,
                conditionProbabilities
        );

        ScoreCalculator.Result result = scoreCalculator.calculate(
                calculatorSelections,
                statWeightsSupplier.get(),
                conditionProbabilities,
                request.getUserStats()
        );
        return toResponse(result, undefinedConditionWarnings);
    }

    private Integer validateBattingOrder(Integer battingOrder) {
        if (battingOrder == null) {
            throw badRequest("Batting order is required for batters.");
        }
        if (battingOrder < 1 || battingOrder > 9) {
            throw badRequest("Batting order must be between 1 and 9.");
        }
        return battingOrder;
    }

    private void validateSelections(List<ScoreSelection> selections, String cardType) {
        if (selections == null || selections.isEmpty()) {
            throw badRequest("At least one skill selection is required.");
        }
        int slotCount = SkillRules.slotCount(cardType);
        if (selections.size() > slotCount) {
            throw badRequest("Too many skill selections for card type " + cardType + ".");
        }
        Set<String> seenSkillIds = new HashSet<>();
        for (ScoreSelection selection : selections) {
            if (selection == null) {
                throw badRequest("Skill selection is required.");
            }
            String skillId = normalizeRequiredOrThrow(selection.getSkillId(), "Skill selection is required.");
            if (!seenSkillIds.add(skillId)) {
                throw badRequest("Duplicate skill selection is not allowed.");
            }
            if (selection.getLevel() == null || selection.getLevel() < 1) {
                throw badRequest("Skill level must be at least 1.");
            }
        }
    }

    private ScoreSkillOption toOption(ScoreSkill skill) {
        return ScoreSkillOption.builder()
                .skillId(skill.getSkillKey())
                .cardType(skill.getCardType())
                .position(skill.getPosition())
                .name(skill.getName())
                .description(skill.getDescription())
                .maxLevel(maxLevel(skill))
                .levelLabels(SkillRules.gradeLabels(skill.getCardType(), maxLevel(skill)))
                .build();
    }

    private List<ScoreSkill> scoreSkillsForCardType(String cardType) {
        return allowedSkillCardTypes(cardType).stream()
                .flatMap(type -> scoreSkillRepository.findByCardTypeIgnoreCase(type).stream())
                .toList();
    }

    private List<String> allowedSkillCardTypes(String cardType) {
        return switch (cardType) {
            case "BLACK" -> List.of("NORMAL", "BLACK");
            case "WBC" -> List.of("NORMAL", "WBC");
            case "WBC_BLACK" -> List.of("NORMAL", "WBC", "BLACK");
            case "MOMENT" -> List.of("NORMAL", "MOMENT");
            case "SUPREME_MOMENT" -> List.of("NORMAL", "MOMENT");
            case "HOF" -> List.of("NORMAL", "HOF");
            default -> List.of(cardType);
        };
    }

    private int maxLevel(ScoreSkill skill) {
        return SkillRules.maxLevel(skill);
    }

    private UndefinedConditionWarnings applyUndefinedConditionWarnings(
            List<ScoreCalculator.Selection> selections,
            Map<String, Double> conditionProbabilities
    ) {
        List<String> globalWarnings = new ArrayList<>();
        Map<String, List<String>> perSkillWarnings = new HashMap<>();
        Set<String> seenWarnings = new HashSet<>();

        for (ScoreCalculator.Selection selection : selections) {
            ScoreSkill skill = selection.skill();
            for (ScoreEffect effect : skill.getEffects()) {
                for (String token : conditionTokens(effect.getCondition())) {
                    if (conditionProbabilities.containsKey(token)) {
                        continue;
                    }
                    conditionProbabilities.put(token, 0.0);
                    String skillKey = skill.getSkillKey();
                    String globalWarning = skillKey + " contains undefined condition: " + token;
                    if (seenWarnings.add(globalWarning)) {
                        globalWarnings.add(globalWarning);
                        perSkillWarnings
                                .computeIfAbsent(skillKey, ignored -> new ArrayList<>())
                                .add("Undefined condition: " + token);
                    }
                }
            }
        }

        return new UndefinedConditionWarnings(globalWarnings, perSkillWarnings);
    }

    private List<String> conditionTokens(String condition) {
        if (condition == null || condition.isBlank() || condition.equalsIgnoreCase("ALWAYS")) {
            return List.of();
        }
        List<String> tokens = new ArrayList<>();
        for (String part : condition.split("\\+")) {
            String token = part.trim();
            if (!token.isEmpty() && !"ALWAYS".equalsIgnoreCase(token)) {
                tokens.add(token);
            }
        }
        return tokens;
    }

    private ScoreResponse toResponse(ScoreCalculator.Result result, UndefinedConditionWarnings undefinedConditionWarnings) {
        return ScoreResponse.builder()
                .total(result.total())
                .perSkill(result.perSkill().stream()
                        .map(skill -> ScoreResponse.SkillScore.builder()
                                .skillId(skill.skillKey())
                                .name(skill.name())
                                .score(skill.score())
                                .perStat(toStatScores(skill.perStat()))
                                .breakdown(skill.breakdown().stream()
                                        .map(this::toEffectBreakdown)
                                        .toList())
                                .warnings(undefinedConditionWarnings.perSkillWarnings()
                                        .getOrDefault(skill.skillKey(), List.of()))
                                .build())
                        .toList())
                .perStat(toStatScores(result.perStat()))
                .warnings(undefinedConditionWarnings.globalWarnings())
                .build();
    }

    private ScoreResponse.EffectBreakdown toEffectBreakdown(ScoreCalculator.EffectBreakdown breakdown) {
        return ScoreResponse.EffectBreakdown.builder()
                .stat(breakdown.stat())
                .condition(breakdown.condition())
                .weight(breakdown.weight())
                .value(breakdown.value())
                .conditionProbability(breakdown.conditionProbability())
                .subtotal(breakdown.subtotal())
                .baseStat(breakdown.baseStat())
                .baseValue(breakdown.baseValue())
                .rawValue(breakdown.rawValue())
                .build();
    }

    private List<ScoreResponse.StatScore> toStatScores(Map<String, Double> perStat) {
        return perStat.entrySet().stream()
                .map(entry -> ScoreResponse.StatScore.builder()
                        .stat(entry.getKey())
                        .value(entry.getValue())
                        .build())
                .toList();
    }

    private String normalizeCardTypeOrThrow(String cardType) {
        try {
            return SkillRules.normalizeCardType(cardType);
        } catch (IllegalArgumentException ex) {
            throw badRequest(ex.getMessage());
        }
    }

    private String normalizeRequiredOrThrow(String value, String message) {
        try {
            return SkillRules.normalizeRequired(value, message);
        } catch (IllegalArgumentException ex) {
            throw badRequest(message);
        }
    }

    private ResponseStatusException badRequest(String message) {
        return new ResponseStatusException(HttpStatus.BAD_REQUEST, message);
    }

    private record UndefinedConditionWarnings(
            List<String> globalWarnings,
            Map<String, List<String>> perSkillWarnings
    ) {
    }

    /**
     * 전체 스킬을 S레벨 기준으로 채점해 티어별 내림차순으로 돌려준다.
     *
     * <p>S가 없는 스킬(수치 단계가 5단계 미만)은 자기 최대 등급으로 내려서 채점하고,
     * 실제 적용된 등급을 함께 반환한다. 효과 행이 없는 스킬은 0점으로 포함한다 —
     * 능력치가 아니라 확률을 바꾸는 효과라 현재 모델로 측정할 수 없을 뿐이다.
     *
     * @param topN 티어별 상위 몇 개까지 담을지. 0 이하면 전부 담는다.
     */
    public ScoreTableResponse buildScoreTable(ScoreTableRequest request, int topN) {
        String normalizedPosition = normalizeRequiredOrThrow(request.getPosition(), "Position selection is required.");
        Map<String, Double> conditionProbabilities = ScoreCalculator.conditionProbabilitiesForPosition(
                normalizedPosition,
                request.getBattingOrder(),
                request.getPitcherSlot(),
                null,
                request.getThrowHand(),
                request.getBatHand()
        );
        Map<String, Double> statWeights = statWeightsSupplier.get();

        Map<SkillTier, List<ScoreTableResponse.Entry>> grouped = new EnumMap<>(SkillTier.class);
        for (ScoreSkill skill : scoreSkillRepository.findAll()) {
            SkillTier tier = SkillTier.of(skill);
            if (tier == null || !SkillRules.matchesPosition(skill.getPosition(), normalizedPosition)) {
                continue;
            }
            int level = appliedLevel(skill);
            ScoreCalculator.Result result = scoreCalculator.calculate(
                    List.of(new ScoreCalculator.Selection(skill, level)),
                    statWeights,
                    conditionProbabilities,
                    request.getUserStats()
            );
            grouped.computeIfAbsent(tier, key -> new ArrayList<>()).add(ScoreTableResponse.Entry.builder()
                    .skillId(skill.getSkillKey())
                    .name(skill.getName())
                    .description(skill.getDescription())
                    .score(result.total())
                    .appliedGrade(SkillRules.gradeLabels(skill.getCardType(), maxLevel(skill)).get(level - 1))
                    .build());
        }

        List<ScoreTableResponse.TierGroup> tiers = new ArrayList<>();
        for (SkillTier tier : SkillTier.DISPLAY_ORDER) {
            List<ScoreTableResponse.Entry> entries = grouped.getOrDefault(tier, List.of());
            if (entries.isEmpty()) {
                continue;
            }
            entries.sort(Comparator.comparingDouble(ScoreTableResponse.Entry::getScore).reversed()
                    .thenComparing(ScoreTableResponse.Entry::getSkillId));
            int total = entries.size();
            tiers.add(ScoreTableResponse.TierGroup.builder()
                    .tier(tier.key())
                    .totalCount(total)
                    .entries(topN > 0 && total > topN ? List.copyOf(entries.subList(0, topN)) : List.copyOf(entries))
                    .build());
        }
        return ScoreTableResponse.builder().tiers(tiers).build();
    }

    /** S레벨. 사다리에서 S 위치가 스킬의 최대 단계를 넘으면 최대 단계로 내린다. */
    private int appliedLevel(ScoreSkill skill) {
        return Math.min(SkillRules.levelIndex(Level.S, skill.getCardType()), maxLevel(skill));
    }
}

File: backend/src/main/kotlin/com/example/skillsim/service/ScoreService.kt
package com.example.skillsim.service

import com.example.skillsim.config.ScoreDataLoader
import com.example.skillsim.dto.ScoreRequest
import com.example.skillsim.dto.ScoreResponse
import com.example.skillsim.dto.ScoreSelection
import com.example.skillsim.dto.ScoreSkillOption
import com.example.skillsim.dto.ScoreTableRequest
import com.example.skillsim.dto.ScoreTableResponse
import com.example.skillsim.enums.Handedness
import com.example.skillsim.enums.Level
import com.example.skillsim.model.ScoreSkill
import com.example.skillsim.repository.ScoreSkillRepository
import java.util.EnumMap
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException

@Service
class ScoreService private constructor(
    private val scoreSkillRepository: ScoreSkillRepository,
    private val scoreCalculator: ScoreCalculator,
    private val statWeightsSupplier: () -> Map<String, Double>,
) {

    @Autowired
    constructor(
        scoreSkillRepository: ScoreSkillRepository,
        scoreCalculator: ScoreCalculator,
        scoreDataLoader: ScoreDataLoader,
    ) : this(scoreSkillRepository, scoreCalculator, scoreDataLoader::statWeights)

    internal constructor(
        scoreSkillRepository: ScoreSkillRepository,
        scoreCalculator: ScoreCalculator,
        statWeights: Map<String, Double>,
    ) : this(scoreSkillRepository, scoreCalculator, { statWeights })

    fun listSkills(cardGrade: String?, cardVariant: String?, position: String?): List<ScoreSkillOption> {
        val card = resolveCard(cardGrade, cardVariant)
        val normalizedPosition = normalizeRequiredOrThrow(position, "Position selection is required.")

        return skillsForCard(card)
            .filter { SkillRules.matchesPosition(it.position, normalizedPosition) }
            .map { it.toOption() }
    }

    fun calculate(request: ScoreRequest?): ScoreResponse {
        if (request == null) {
            throw badRequest("Score request is required.")
        }

        val card = resolveCard(request.cardGrade ?: request.cardType, request.cardVariant)
        val normalizedPosition = normalizeRequiredOrThrow(request.position, "Position selection is required.")
        val selections = request.selections
        validateSelections(selections, card)

        val role = SkillRules.roleForPosition(normalizedPosition)
        val pitcherSlot = request.pitcherSlot
        var battingOrder: Int? = null
        when (role) {
            "BATTER" -> battingOrder = validateBattingOrder(request.battingOrder)
            "SP" -> if (pitcherSlot == null || pitcherSlot !in 1..5) {
                throw badRequest("Pitcher slot must be between 1 and 5 for starting pitchers.")
            }

            "RP" -> if (pitcherSlot == null || pitcherSlot !in 1..6) {
                throw badRequest("Pitcher slot must be between 1 and 6 for relief pitchers.")
            }
        }

        val seenSkillIds = mutableSetOf<String>()
        val poolCounts = mutableMapOf<String, Int>()
        val calculatorSelections = mutableListOf<ScoreCalculator.Selection>()

        selections.orEmpty().forEachIndexed { slotIndex, selection ->
            val skillId = normalizeRequiredOrThrow(selection.skillId, "Skill selection is required.")
            if (!seenSkillIds.add(skillId)) {
                throw badRequest("Duplicate skill selection is not allowed.")
            }

            val skill = scoreSkillRepository.findBySkillKey(skillId)
                ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Score skill not found: $skillId")
            val pool = SkillRules.normalizeSkillPool(skill.cardType)
            if (pool !in CardRules.skillPools(card.grade, card.variant)) {
                throw badRequest("Selected skill does not match requested card grade.")
            }
            // 등장 확률이 0인 자리는 막는다. 모먼트 전용은 첫 칸에만 나오고 블랙은 카드당 한 장이다.
            if (!CardRules.allowsPoolInSlot(pool, slotIndex)) {
                throw badRequest("Moment skills can only be in the first slot.")
            }
            poolCounts[pool] = (poolCounts[pool] ?: 0) + 1
            if (poolCounts.getValue(pool) > CardRules.maxSkillsFromPool(pool)) {
                throw badRequest("A card can have at most ${CardRules.maxSkillsFromPool(pool)} $pool skill.")
            }
            if (!SkillRules.matchesPosition(skill.position, normalizedPosition)) {
                throw badRequest("Selected skill does not match requested position.")
            }
            val maxLevel = SkillRules.maxLevel(skill)
            val level = selection.level
            if (level == null || level !in 1..maxLevel) {
                throw badRequest("Skill level must be between 1 and $maxLevel.")
            }

            calculatorSelections += ScoreCalculator.Selection(skill, level)
        }

        return scoreSelections(
            selections = calculatorSelections,
            position = normalizedPosition,
            cardType = card.grade,
            battingOrder = battingOrder,
            pitcherSlot = pitcherSlot,
            throwHand = request.throwHand,
            batHand = request.batHand,
            userStats = request.userStats,
        )
    }

    /**
     * 검증이 끝난 스킬 선택을 채점한다. [calculate]의 뒷부분이며 덱 채점이 같은 경로를 타도록 뽑아냈다.
     *
     * 검증을 하지 않으므로 호출자가 카드 등급·포지션·레벨을 이미 확인했어야 한다.
     * [cardType]에는 카드 **등급**을 넘긴다(상대등급우세 조건이 이 값을 본다).
     * [battingOrder]가 null이면 [ScoreCalculator]의 기본 타순이 쓰인다 — 타순이 없는
     * 후보 선수를 채점할 때 필요하다.
     */
    internal fun scoreSelections(
        selections: List<ScoreCalculator.Selection>,
        position: String,
        cardType: String,
        battingOrder: Int? = null,
        pitcherSlot: Int? = null,
        throwHand: Handedness? = null,
        batHand: Handedness? = null,
        userStats: Map<String, Double>? = null,
        statBonus: Double = 0.0,
    ): ScoreResponse {
        val conditionProbabilities = ScoreCalculator.conditionProbabilitiesForPosition(
            position = position,
            battingOrder = battingOrder,
            pitcherSlot = pitcherSlot,
            cardType = cardType,
            throwHand = throwHand,
            batHand = batHand,
        )
        val undefinedConditionWarnings =
            applyUndefinedConditionWarnings(selections, conditionProbabilities)

        val result = scoreCalculator.calculate(
            selections,
            statWeightsSupplier(),
            conditionProbabilities,
            userStats,
            statBonus,
        )
        return toResponse(result, undefinedConditionWarnings, selections)
    }

    private fun validateBattingOrder(battingOrder: Int?): Int {
        if (battingOrder == null) {
            throw badRequest("Batting order is required for batters.")
        }
        if (battingOrder !in 1..9) {
            throw badRequest("Batting order must be between 1 and 9.")
        }
        return battingOrder
    }

    private fun validateSelections(selections: List<ScoreSelection>?, card: Card) {
        if (selections.isNullOrEmpty()) {
            throw badRequest("At least one skill selection is required.")
        }
        if (selections.size > CardRules.slotCount(card.grade)) {
            throw badRequest("Too many skill selections for card grade ${card.grade}.")
        }
        val seenSkillIds = mutableSetOf<String>()
        for (selection in selections) {
            val skillId = normalizeRequiredOrThrow(selection.skillId, "Skill selection is required.")
            if (!seenSkillIds.add(skillId)) {
                throw badRequest("Duplicate skill selection is not allowed.")
            }
            val level = selection.level
            if (level == null || level < 1) {
                throw badRequest("Skill level must be at least 1.")
            }
        }
    }

    private fun ScoreSkill.toOption(): ScoreSkillOption {
        val maxLevel = SkillRules.maxLevel(this)
        return ScoreSkillOption(
            skillId = skillKey,
            cardType = cardType,
            position = position,
            name = name,
            description = description,
            maxLevel = maxLevel,
            levelLabels = SkillRules.gradeLabels(cardType, maxLevel),
        )
    }

    private fun skillsForCard(card: Card): List<ScoreSkill> =
        CardRules.skillPools(card.grade, card.variant)
            .flatMap { scoreSkillRepository.findByCardTypeIgnoreCase(it) }

    /** 이 카드가 고를 수 있는 스킬 풀. 등급과 변형 두 축이 정한다([CardRules.skillPools]). */
    internal fun skillPoolsFor(cardGrade: String?, cardVariant: String?): List<String> {
        val card = resolveCard(cardGrade, cardVariant)
        return CardRules.skillPools(card.grade, card.variant)
    }

    private fun applyUndefinedConditionWarnings(
        selections: List<ScoreCalculator.Selection>,
        conditionProbabilities: MutableMap<String, Double>,
    ): UndefinedConditionWarnings {
        val globalWarnings = mutableListOf<String>()
        val perSkillWarnings = mutableMapOf<String, MutableList<String>>()
        val seenWarnings = mutableSetOf<String>()

        for (selection in selections) {
            val skill = selection.skill
            for (effect in skill.effects) {
                for (token in conditionTokens(effect.condition)) {
                    if (conditionProbabilities.containsKey(token)) {
                        continue
                    }
                    conditionProbabilities[token] = 0.0
                    val globalWarning = "${skill.skillKey} contains undefined condition: $token"
                    if (seenWarnings.add(globalWarning)) {
                        globalWarnings += globalWarning
                        perSkillWarnings.getOrPut(skill.skillKey) { mutableListOf() } +=
                            "Undefined condition: $token"
                    }
                }
            }
        }

        return UndefinedConditionWarnings(globalWarnings, perSkillWarnings)
    }

    private fun conditionTokens(condition: String?): List<String> {
        if (condition.isNullOrBlank() || condition.equals("ALWAYS", ignoreCase = true)) {
            return emptyList()
        }
        return condition.split("+")
            .map { it.trim() }
            .filter { it.isNotEmpty() && !it.equals("ALWAYS", ignoreCase = true) }
    }

    private fun toResponse(
        result: ScoreCalculator.Result,
        undefinedConditionWarnings: UndefinedConditionWarnings,
        selections: List<ScoreCalculator.Selection>,
    ): ScoreResponse {
        // 설명의 수치는 사용자가 그 슬롯에서 고른 레벨을 따라야 화면의 점수와 같은 기준이 된다.
        val bySkillKey = mutableMapOf<String, ScoreCalculator.Selection>()
        for (selection in selections) {
            bySkillKey.putIfAbsent(selection.skill.skillKey, selection)
        }
        return ScoreResponse(
            total = result.total,
            perSkill = result.perSkill.map { skill ->
                ScoreResponse.SkillScore(
                    skillId = skill.skillKey,
                    name = skill.name,
                    resolvedDescription = bySkillKey[skill.skillKey]?.let {
                        SkillDescriptions.resolve(it.skill, it.level)
                    },
                    score = skill.score,
                    perStat = skill.perStat.toStatScores(),
                    breakdown = skill.breakdown.map { it.toEffectBreakdown() },
                    warnings = undefinedConditionWarnings.perSkillWarnings[skill.skillKey].orEmpty(),
                )
            },
            perStat = result.perStat.toStatScores(),
            warnings = undefinedConditionWarnings.globalWarnings,
        )
    }

    private fun ScoreCalculator.EffectBreakdown.toEffectBreakdown() = ScoreResponse.EffectBreakdown(
        stat = stat,
        condition = condition,
        weight = weight,
        value = value,
        conditionProbability = conditionProbability,
        subtotal = subtotal,
        baseStat = baseStat,
        baseValue = baseValue,
        rawValue = rawValue,
    )

    private fun Map<String, Double>.toStatScores(): List<ScoreResponse.StatScore> =
        map { (stat, value) -> ScoreResponse.StatScore(stat, value) }

    /** 카드 한 장의 두 축. 예전 단일 cardType으로 들어와도 여기서 풀린다. */
    internal data class Card(val grade: String, val variant: String)

    private fun resolveCard(cardGrade: String?, cardVariant: String?): Card = try {
        val grade = CardRules.normalizeGrade(cardGrade)
        // 변형을 따로 주지 않았는데 등급 자리에 예전 이름(WBC_BLACK 등)이 왔다면 거기서 꺼낸다.
        val variant = if (cardVariant.isNullOrBlank()) {
            CardRules.fromLegacyCardType(cardGrade)?.second ?: CardRules.VARIANT_NONE
        } else {
            CardRules.normalizeVariant(cardVariant)
        }
        CardRules.validateCombination(grade, variant)
        Card(grade, variant)
    } catch (ex: IllegalArgumentException) {
        throw badRequest(ex.message ?: "Card grade is required.")
    }

    private fun normalizeRequiredOrThrow(value: String?, message: String): String =
        try {
            SkillRules.normalizeRequired(value, message)
        } catch (ex: IllegalArgumentException) {
            throw badRequest(message)
        }

    private fun badRequest(message: String) = ResponseStatusException(HttpStatus.BAD_REQUEST, message)

    private data class UndefinedConditionWarnings(
        val globalWarnings: List<String>,
        val perSkillWarnings: Map<String, List<String>>,
    )

    /**
     * 전체 스킬을 S레벨 기준으로 채점해 티어별 내림차순으로 돌려준다.
     *
     * S가 없는 스킬(수치 단계가 5단계 미만)은 자기 최대 등급으로 내려서 채점하고,
     * 실제 적용된 등급을 함께 반환한다. 효과 행이 없는 스킬은 0점으로 포함한다 —
     * 능력치가 아니라 확률을 바꾸는 효과라 현재 모델로 측정할 수 없을 뿐이다.
     *
     * @param topN 티어별 상위 몇 개까지 담을지. 0 이하면 전부 담는다.
     */
    fun buildScoreTable(request: ScoreTableRequest, topN: Int): ScoreTableResponse {
        val normalizedPosition = normalizeRequiredOrThrow(request.position, "Position selection is required.")
        val conditionProbabilities = ScoreCalculator.conditionProbabilitiesForPosition(
            position = normalizedPosition,
            battingOrder = request.battingOrder,
            pitcherSlot = request.pitcherSlot,
            cardType = null,
            throwHand = request.throwHand,
            batHand = request.batHand,
        )
        val statWeights = statWeightsSupplier()

        val grouped = EnumMap<SkillTier, MutableList<ScoreTableResponse.Entry>>(SkillTier::class.java)
        for (skill in scoreSkillRepository.findAll()) {
            val tier = SkillTier.of(skill)
            if (tier == null || !SkillRules.matchesPosition(skill.position, normalizedPosition)) {
                continue
            }
            val level = appliedLevel(skill)
            val result = scoreCalculator.calculate(
                listOf(ScoreCalculator.Selection(skill, level)),
                statWeights,
                conditionProbabilities,
                request.userStats,
            )
            grouped.getOrPut(tier) { mutableListOf() } += ScoreTableResponse.Entry(
                skillId = skill.skillKey,
                name = skill.name,
                description = skill.description,
                resolvedDescription = SkillDescriptions.resolve(skill, level),
                score = result.total,
                appliedGrade = SkillRules.gradeLabels(skill.cardType, SkillRules.maxLevel(skill))[level - 1],
            )
        }

        val tiers = SkillTier.DISPLAY_ORDER.mapNotNull { tier ->
            val entries = grouped[tier].orEmpty()
            if (entries.isEmpty()) {
                return@mapNotNull null
            }
            val sorted = entries.sortedWith(
                compareByDescending<ScoreTableResponse.Entry> { it.score }.thenBy { it.skillId },
            )
            ScoreTableResponse.TierGroup(
                tier = tier.key,
                totalCount = sorted.size,
                entries = if (topN > 0 && sorted.size > topN) sorted.take(topN) else sorted,
            )
        }
        return ScoreTableResponse(tiers)
    }

    /** S레벨. 사다리에서 S 위치가 스킬의 최대 단계를 넘으면 최대 단계로 내린다. */
    private fun appliedLevel(skill: ScoreSkill): Int =
        minOf(SkillRules.levelIndex(Level.S, skill.cardType), SkillRules.maxLevel(skill))
}
